ownCloud/Client/Tasks/ScheduledTaskManager.swift

```diff
@@ -0,0 +1,243 @@
+			wifiMonitor = NWPathMonitor(requiredInterfaceType: .wifi)
+			(wifiMonitor as? NWPathMonitor)?.pathUpdateHandler = { [weak self] path in
+				// Use "inexpensive" WiFi only (not behind a cellular hot-spot)
+				let wifiAvailable =  (path.status == .satisfied && !path.isExpensive)
+				self?.wifiDetected = wifiAvailable
+				if wifiAvailable {
+					self?.scheduleTasks()
+				}
+			}
+			(wifiMonitor as? NWPathMonitor)?.start(queue: wifiMonitorQueue!)
+		}
+
+		checkPowerState()
+
+		startMonitoringPhotoLibraryChangesIfNecessary()
+	}
+
+	deinit {
+		NotificationCenter.default.removeObserver(self)
+		if #available(iOS 12, *) {
+			(wifiMonitor as? NWPathMonitor)?.cancel()
+		}
+		stopMonitoringPhotoLibraryChanges()
+	}
+
+	// MARK: - Notifications handling
+
+	@objc private func applicationStateChange(notificaton:Notification) {
+		switch notificaton.name {
+		case UIApplication.didBecomeActiveNotification:
+			state = .foreground
+		case UIApplication.didEnterBackgroundNotification:
+			state = .background
+		default:
+			break
+		}
+
+		scheduleTasks()
+	}
+
+	@objc private func batteryLevelDidChange(notification:Notification) {
+		checkPowerState()
+		if lowBatteryDetected || externalPowerConnected {
+			scheduleTasks()
+		}
+	}
+
+	@objc private func mediaUploadSettingsDidChange(notification:Notification) {
+		if shallMonitorPhotoLibraryChanges() {
+			startMonitoringPhotoLibraryChangesIfNecessary()
+		} else {
+			stopMonitoringPhotoLibraryChanges()
+		}
+	}
+
+	// MARK: - Background fetching
+
+	func backgroundFetch(completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
+		self.state = .backgroundFetch
+		scheduleTasks(fetchCompletion: completionHandler)
+	}
+
+	// MARK: - Private methods
+
+	private func scheduleTasks(fetchCompletion:((UIBackgroundFetchResult) -> Void)? = nil, completion:((_ scheduledTaskCount:Int)->Void)? = nil) {
+		OnMainThread {
+			// Build a context
+			let location = OCExtensionLocation(ofType: .scheduledTask, identifier: self.state.locationIdentifier())
+
+			// Add requirements
+			var requirements = [String : Bool]()
+			if self.wifiDetected {
+				requirements[ScheduledTaskAction.FeatureKeys.runOnWifi] = true
+			}
+			if self.lowBatteryDetected {
+				requirements[ScheduledTaskAction.FeatureKeys.runOnLowBattery] = true
+			}
+			if self.externalPowerConnected {
+				requirements[ScheduledTaskAction.FeatureKeys.runOnExternalPower] = true
+			}
+			if self.photoLibraryChangeDetected {
+				requirements[ScheduledTaskAction.FeatureKeys.photoLibraryChanged] = true
+			}
+
+			let context = OCExtensionContext(location: location, requirements: requirements, preferences: nil)
+
+			// Find a task to run
+			if let matches = try? OCExtensionManager.shared.provideExtensions(for: context) {
+				for match in matches {
+					if let task = match.extension.provideObject(for: context) as? ScheduledTaskAction {
+						if self.state == .backgroundFetch {
+							task.backgroundFetchCompletion = fetchCompletion
```

Comment:
With `task.backgroundFetchCompletion` called in `task.completion()`, the `fetchCompletion` block will be called multiple times (and possibly prematurely) if there's more than one extension running.

AFAIK the `fetchCompletion` block provided by iOS should only called once - after all tasks have finished.

Reply:
@felix-schwarz fixed by moving the logic calling OS fetch completion block to the ScheduledTaskManager itself which is called when completion handlers of all background-fetch tasks have been called.